### PR TITLE
feat: enable admin key to be able to remove other keys

### PR DESCRIPTION
**Pull request by fchitakova**
**Description**:

The purpose of these changes is to enable the admin key to be able remove itself from a token, as well as the other low privileged keys as described in [HIP-540](https://hips.hedera.com/hip/hip-540  ). 

To achieve this, the change checks:

1. If the transaction body contains IMMUTABILITY_SENTINEL_KEY value for the any key, since this is how we expect the keys to be removed. 
2. If the admin keys is present and used for signing the transaction in addition to the required payer signature

**Related issue(s)**: https://github.com/hashgraph/hedera-services/issues/12158

Fixes #

**Notes for reviewer**:

**Checklist**

- [X] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

File: hedera-node/hedera-app/src/xtest/java/token/TokenUpdateXTest.java [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+                                .token(idOfNamedToken(TOKEN_ID))
+                                .adminKey(IMMUTABILITY_SENTINEL_KEY))
+                        .build(),
+                OK);
+    }
+
+    @Override
+    protected Map<TokenID, Token> initialTokens() {
+        /**
+         * Initial token key must always have initialized all keys. IMMUTABILITY_SENTINEL_KEY is the default value for keys.
```
Comment (petreze) [inline hedera-node/hedera-app/src/xtest/java/token/TokenUpdateXTest.java:134]: `IMMUTABILITY_SENTINEL_KEY` is not the default value for keys, it is the key value to consider the key as "_removed_"
could be changed to "Initial token must always have initialized all keys. Set DEFAULT_KEY as default value for all keys."

**Review by petreze — COMMENTED** [review]
Still WIP, suites need to be added as well

File: hedera-node/hedera-app/src/xtest/java/token/TokenUpdateXTest.java (hunk 1 above)
Comment (fchitakova) [reply hedera-node/hedera-app/src/xtest/java/token/TokenUpdateXTest.java]: Updated, thanks!

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java [hunk 2]
```diff
@@ -399,15 +400,48 @@ private void updateKeys(
...
-                builder.adminKey((Key) null);
-            } else {
+            if (!isKeyRemoval(newAdminKey)) {
                 builder.adminKey(newAdminKey);
             }
         }
     }
 
+    private void removeKeysIfNeeded(final TokenUpdateTransactionBody op, final Token.Builder builder) {
+        if (IMMUTABILITY_SENTINEL_KEY.equals(op.adminKey())) {
```
Comment (petreze) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java:412]: all occurences of `IMMUTABILITY_SENTINEL_KEY.equals(*some_key*)` in `removeKeysIfNeeded` can be replaced with `isKeyRemoval(*some_key*)`

File: hedera-node/hedera-app/src/xtest/java/token/TokenUpdateXTest.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+                OK);
+        handleAndCommitSingleTransaction(
+                component.tokenUpdateHandler(),
+                TransactionBody.newBuilder()
+                        .transactionID(TransactionID.newBuilder()
+                                .accountID(DEFAULT_PAYER_ID)
+                                .build())
+                        .tokenUpdate(TokenUpdateTransactionBody.newBuilder()
+                                .token(idOfNamedToken(TOKEN_ID))
+                                .metadataKey(IMMUTABILITY_SENTINEL_KEY))
```
Comment (petreze) [inline hedera-node/hedera-app/src/xtest/java/token/TokenUpdateXTest.java:115]: In the HIP there are **no** specifications about the `metadataKey`, hence all 
`metadataKey` mentions are irrelevant and should be removed. This includes all verification mentions and changes

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java (hunk 2 above)
Comment (fchitakova) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java]: Added, thanks

File: hedera-node/hedera-app/src/xtest/java/token/TokenUpdateXTest.java (hunk 3 above)
Comment (fchitakova) [reply hedera-node/hedera-app/src/xtest/java/token/TokenUpdateXTest.java]: Removed, thanks

File: hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/queries/token/HapiGetTokenInfo.java [hunk 4]
```diff
@@ -89,6 +90,17 @@ public HapiGetTokenInfo(String token) {
...
+
+    private boolean emptyAdminKeyList = false;
+
+    private boolean emptyFreezeKey = false;
+
+    private boolean emptyKycKey = true;
+
+    private boolean emptySupplyKey = true;
+
+    private boolean emptyWipeKey = true;
```
Comment (petreze) [inline hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/queries/token/HapiGetTokenInfo.java:103]: I suppose `emptyKycKey`, `emptySupplyKey` and `emptyWipeKey` has to be also `false` as default values

**Review by petreze — DISMISSED** [review]
LGTM!

**Review by petreze — APPROVED** [review]

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java [hunk 5]
```diff
@@ -399,15 +400,48 @@ private void updateKeys(
...
+        if (isKeyRemoval(op.feeScheduleKey())) {
+            builder.feeScheduleKey((Key) null);
+        }
+
+        if (isKeyRemoval(op.pauseKey())) {
+            builder.pauseKey((Key) null);
+        }
+    }
+
+    private boolean isKeyRemoval(Key key) {
```
[the remark this answers is not in the document]
Comment (fchitakova) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java:441]: @mustafauzunn , the problem is that we do not want to have a null check

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java [hunk 6]
```diff
@@ -408,6 +436,39 @@ private void updateKeys(
...
+     * @param context pre handle context
+     * @param op token update transaction body
+     * @param originalToken original token
+     */
+    private void addRequiredSignersForKeyRemoval(
+            @NonNull PreHandleContext context,
+            @NonNull final TokenUpdateTransactionBody op,
+            @NonNull final Token originalToken)
+            throws PreCheckException {
+        if (op.hasAdminKey()) {
```
[the remark this answers is not in the document]
Comment (petreze) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/TokenUpdateHandler.java:451]: Yup, currently I'm working on all the conflicts that appeared! There are heavy changes that mess with this logic and some test failures.

On the other topic we will not require the `adminKey` to sign in the cases where a low priority key wants to update itself as well as the case where the `metadataKey` wants to update the `metadata` field

@netopyr

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/BaseTokenHandler.java [hunk 7]
```diff
@@ -427,6 +428,52 @@ public static boolean isMetadataOnlyUpdateOp(@NonNull final TokenUpdateTransacti
...
+    public static boolean isAdminOrLowPriorityKeyRemoval(
+            @NonNull final TokenUpdateTransactionBody op, @NonNull final Token originalToken) {
+        return originalToken.hasAdminKey() && IMMUTABILITY_SENTINEL_KEY.equals(op.adminKey())
+                || originalToken.hasWipeKey() && IMMUTABILITY_SENTINEL_KEY.equals(op.wipeKey())
+                || originalToken.hasKycKey() && IMMUTABILITY_SENTINEL_KEY.equals(op.kycKey())
+                || originalToken.hasSupplyKey() && IMMUTABILITY_SENTINEL_KEY.equals(op.supplyKey())
+                || originalToken.hasFreezeKey() && IMMUTABILITY_SENTINEL_KEY.equals(op.freezeKey())
+                || originalToken.hasFeeScheduleKey() && IMMUTABILITY_SENTINEL_KEY.equals(op.feeScheduleKey())
+                || originalToken.hasMetadataKey() && IMMUTABILITY_SENTINEL_KEY.equals(op.pauseKey())
+                || originalToken.hasMetadataKey() && IMMUTABILITY_SENTINEL_KEY.equals(op.metadataKey());
```
Comment (petreze) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/handlers/BaseTokenHandler.java:444]: Checks whether `originalToken` has some key are not necessary because [here](https://github.com/hashgraph/hedera-services/blob/12158-enable-admin-key-to-remove-all-keys/hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/validators/TokenUpdateValidator.java#L65) we make them and will indicate with `TOKEN_IS_IMMUTABLE`. We can replace them with checks if `op` has some key to ensure it is not a `null`.
Also `IMMUTABILITY_SENTINEL_KEY.equals(someKey)` can be replaced with `isKeyRemoval(someKey)`

_All of these apply to all key occurrences_

**Comment by petreze** [comment]
Closing this one as we created a [PR](https://github.com/hashgraph/hedera-services/pull/12736) for the entire implementation. This was made because there are quite a lot corner cases which test the entire logic of the HIP
